obs-mcp/pkg/prometheus/guardrails.go

```diff
@@ -102,27 +119,155 @@ func (g *Guardrails) IsSafeQuery(query string) (bool, error) {
 				// Rule 3: Check for expensive regex matchers on *any* label i.e blanket matchers
 				if g.DisallowBlanketRegex {
 					isRegex := m.Type == labels.MatchRegexp || m.Type == labels.MatchNotRegexp
-					if isRegex {
-						if m.Value == ".*" || m.Value == ".+" {
-							foundUnsafe = true
-							return fmt.Errorf("unsafe")
+					if isRegex && (m.Value == ".*" || m.Value == ".+") {
+						// If MaxLabelCardinality is set, defer the check to TSDB lookup
+						if g.MaxLabelCardinality > 0 {
+							hasBlanketRegex = true
+							blanketRegexLabels = append(blanketRegexLabels, m.Name)
+						} else {
+							// MaxLabelCardinality is 0, always disallow blanket regex here
+							unsafeReason = fmt.Errorf("query uses blanket regex (%s) on label %q, which is disallowed", m.Value, m.Name)
+							return unsafeReason
 						}
 					}
 				}
 			}
 
 			// Rule 2: All vector selectors must have at least one non-name label matcher
 			if g.RequireLabelMatcher && !hasNonNameMatcher {
-				foundUnsafe = true
-				return fmt.Errorf("unsafe")
+				metricName := n.Name
+				if metricName == "" {
+					metricName = "<metric>"
+				}
+				unsafeReason = fmt.Errorf("query for metric %q does not have any label matchers, which is required", metricName)
+				return unsafeReason
 			}
 		}
 		return nil
 	})
 
-	if foundUnsafe {
-		return false, nil
+	if unsafeReason != nil {
+		return false, unsafeReason
+	}
+
+	// If blanket regex was found but we can't check TSDB (no client), reject the query
+	// this is only set to true when MaxLabelCardinality is set and DisallowBlanketRegex is true
+	if hasBlanketRegex && (client == nil || ctx == nil) {
+		return false, fmt.Errorf("query uses blanket regex on labels %v, but cannot verify label cardinality without TSDB client", blanketRegexLabels)
+	}
```

Comment:
I see that the context here comes from the tool handler's registration and passed down. But in the absence of one, I'd reckon we'd want to surface that in the error as well?